### PR TITLE
When too many refs, allow users to search and cycle through branches and tags

### PR DESCRIPTION
**Pull request by codingtwinky**
fix: #973

* Will always display ref search as long as each nodes has at least one ref available
* Selecting hidden ref will bring it to the left most order and pop off the last displayed ref
* Selecting already displayed ref will bring existing ref to the left most order
* Number of refs to display is configurable by `numRefsToShow`
  * Distribution between branches and tags is not straight forward.  2/5 of the `numRefsToShow` will be used to show tags and `numRefsToShow` - ${length of the tags} will be used to show branches.  This means if `numRefsToShow` is 5 and a git node has 0 tags, ungit will display 5 branches no matter how many branches exists.  If `numRefsToShow` is 5 and git node has 1 tag, ungit will display 4 branches no matter how many branches exists.

### CODE REVIEW

File: components/graph/graph.html [hunk 1]
```diff
@@ -43,7 +43,7 @@
         <!-- /ko -->
 
         <!-- ko if: showNewRefAction -->
-        <span class="newRef" data-bind="css: { editing: branchingFormVisible }">
+        <span class="newRef" data-bind="click: function() {console.log(3333333)}, css: { editing: branchingFormVisible }">
```
Comment (campersau) [inline components/graph/graph.html]: Remove `console.log`

File: components/graph/graph.less [hunk 2]
```diff
@@ -167,6 +167,20 @@
         padding: 0px;
         margin: 0px;
         margin-top: 9px;
+        -webkit-text-stroke-width: 1px;
+        color: rgba(255, 255, 255, 0.3);
+        &:hover {
+          color: rgba(255, 255, 255, 0.8);
+        }
+      }
+      .showSearchForm{
```
Comment (campersau) [inline components/graph/graph.less]: Nit: add space

File: components/graph/git-node.js [hunk 3]
```diff
@@ -62,13 +89,21 @@ var GitNodeViewModel = function(graph, sha1) {
...
   });
   this.branchOrder = ko.observable();
   this.aboveNode = undefined;
   this.belowNode = undefined;
+  this.refSearchFormVisible = ko.observable(false);
+  this.refSearchFormVisible.subscribe((newValue) => {
+    if (!newValue) {
+      // Small timeout because in ff the form is hidden before the submit click event is registered otherwise
+      setTimeout(() => {
+        this.refSearchFormVisible(false);
```
Comment (campersau) [inline components/graph/git-node.js]: I tested it in Firefox too. But for me it worked also without this code.
Also I don't quite understand why setting `refSearchFormVisible` to `false` after it is already set to `false` should do anything. Doesn't knockout check if the value is the same as before and does nothing?

File: components/graph/git-node.js [hunk 4]
```diff
@@ -24,12 +27,33 @@ var GitNodeViewModel = function(graph, sha1) {
...
+    if (newValue) {
+      this.branches(newValue.filter((r) => r.isBranch));
+      this.tags(newValue.filter((r) => r.isTag));
+      this.tagsToDisplay(this.tags.slice(0, maxTagsToDisplay));
+      this.branchesToDisplay(this.branches.slice(0, ungit.config.numRefsToShow - this.tagsToDisplay().length));
+    } else {
+      this.branches.removeAll();
+      this.tags.removeAll();
+      this.branchesToDisplay.removeAll();
+      this.tags.removeAll();
```
Comment (campersau) [inline components/graph/git-node.js]: Too late but I think this should be `this.tagsToDisplay.removeAll();`.

Comment (codingtwinky) [reply components/graph/git-node.js]: thanks for the catch, will submit one soon if no one has

**Pull request merged**